Add burningman accounting

HenrikJannsen opened the pull request:
Implements https://github.com/bisq-network/proposals/issues/390#issuecomment-1319467686

HenrikJannsen committed:
Add allowEmpty flag to InputValidator and BtcAddressValidator

We will use that for the optional address field in compensation requests and interpret empty input as Optional.empty.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add burningManReceiverAddress as extraMap entry to CompensationProposal.
We cannot add a new field as that would break DAO consensus.

Add optional text field for burningManReceiverAddress to CompensationProposal UI.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add new methods to DaoStateService and DaoFacade.

Will be used later by BurningMan domain.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add BurningMan domain

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add new addTableViewWithHeaderAndFilterField method to FormBuilder (will be used later)
Change return type of FormBuilder.addTableViewWithHeader

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add Burningman UI

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add burningManSelectionHeight fields to OfferAvailabilityRequest, InputsForDepositTxRequest and ProcessModel.
Use BurningManService.getBtcFeeReceiverAddress at createFeeTx tasks.
Verify takers burningManSelectionHeight if BurningManService.isActivated at MakerProcessesInputsForDepositTxRequest and OpenOfferManager.
Stores burningManSelectionHeight in ProcessModel.
Add BurningManService fields to OfferAvailabilityModel, PlaceOfferModel, OpenOfferManager and Provider

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Remove FeeReceiverSelector and test as not used anymore

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Use Burningmen receivers for delayed payout tx.
Add verification by re-creating receivers and check txId of both DPTs.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Validate delayedPayoutTx outputs size and amount only if BurningManService.isActivated is false.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add verifyDelayedPayoutTxReceivers method in RefundManager.
Add burningManSelectionHeight and tradeTxFee in Dispute.
Call validateDonationAddressMatchesAnyPastParamValues and validateDonationAddress
only if legacy BM was used.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Fix handling of null entries in getBtcFeeReceiverAddress method

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Make codacy happy

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Sort getBtcFeeReceiverAddress by amount and address.
Increase minOutputAmount to 1000 sat.
Cleanups

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add expectedRevenue to BurningManCandidate.
Add getAverageDistributionPerCycle method to BurningManService.
Show receiver address when BM is selected.
Refactor code, cleanups, UI improvements.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Subtract accumulatedDecayedBurnAmount from allowedBurnAmount

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Cleanup

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Refactor BurningManService

Split up into 4 service classes
- BurningManService: Common stuff
- BurningManInfoService: For displaying BurningMan data
- BtcFeeReceiverService: For getting btcFeeReceivers
- DelayedPayoutTxReceiverService: For getting delayedPayoutTxReceivers. This is the critical part where we need to have a deterministic data and which could break trade consensus.

WIP refactoring. More to come...

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Move getAverageDistributionPerCycle to BurningManService and use from BurningManInfoService.
Remove reference to BurningManInfoService inside BurningManService.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Remove reference to delayedPayoutTxReceiverService and btcFeeReceiverService inside BurningManService.
Use DelayedPayoutTxReceiverService and BtcFeeReceiverService in client code instead of BurningManService.
Make BurningManService package private.

HenrikJannsen committed:
Remove unused fields

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add sorting for output index needed for genesis tx outputs.
Use static factory methods for the 2 diff. use cases of CompensationModel.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Split up calculateShare into 4 distinct methods
Reduce visibility to package scope

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Rearrange methods

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Move BurningManUtil.getLegacyBurningManAddress to BurningManService.
Remove BurningManUtil

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Rename getBtcFeeReceiverAddress to getAddress
Cleanup

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Remove `Current` from method names

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Rename getDelayedPayoutTxReceivers to getReceivers
Cleanup

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Refactor out fields from BurningManCandidate which are not affecting data for delayed payout tx receivers.
This allows to split off the burn target related code to separate classes.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Fix incorrect cycle index

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Use months and not cycles for burn target and average.
Add checks for chainHeight to make code more generic (in the current usage its always last block anyway)

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add applyBlock method and call at constructor (some classes get not initiated at app startup but created by UI activity).

Remove accumulatedDecayedBurnAmount at getAllowedBurnAmount. Was incorrect.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add balance listener for burn bsq input

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Fix wrong initial value.

For myGenesisOutputNames we check for null not Optional.empty as empty is a valid case and we do not
want to wrap the optional field into another Optional.
Also myCompensationRequestNames is using a null check and gets reset to null instead of an
empty set as the empty set is a valid case as well.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add option to add opReturn data at BSQ wallet send screen for non-BSQ balance.

This can be used for providing proof of ownership of a genesis output when attaching a custom BM receiver address.
See: https://github.com/bisq-network/proposals/issues/390#issuecomment-1306075295

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add isDevMode check to set dev params

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add isDevTesting method for adjusting behaviour for dev testing purposes

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Bugfix: set burningManSelectionHeight and tradeTxFee at dispute clone

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add details if donationAddressOfDelayedPayoutTx check fails

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Use null as default value for donationAddressString

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Change ISSUANCE_BOOST_FACTOR to 3
Use lower dev test value for BURN_TARGET_BOOST_AMOUNT

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add util class for cycle related methods with caching
Add getCycleAtIndex convenience method to DaoStateService

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Use cycles for periods instead of blocks in months/years.
A cycle is the fundamental unit in the DAO, months/years have no relevance, so using
cycles as base helps avoids some potential weird behaviour with data getting updated at cycle events.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Avoid negative cycle index

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Fix duration calculations

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Improve burn target calculation and display

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Fix burn target calculations

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add legacy burningman to btc fee distribution in case we do not reach 100% from capped burn shares.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add logs for delayedPayoutTx receiver data

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Fix getSnapshotHeight method

For chainHeights < genesisHeight+3*grid we ended up with future chainHeights

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add LegacyBurningMan to UI

It is not included in BurningManCandidate candidate map as that would complicate things.
We only want to show it in the UI for informational purposes.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Don't show decayedAmount for LegacyBurningMan

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Use TradeWalletService.MIN_DELAYED_PAYOUT_TX_FEE as min value for miner fee in DelayedPayoutTxReceiverService.getSpendableAmount

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Remove unused methods

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Refactoring: Move BM model classes into model package. Adjust visibility.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Refactoring: Move isActivated method to BurningManService

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Use check for isRegtest instead of test for isDevTesting to set dev parameters

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Cleanup/add comments

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Split up legacy BM into DPT and BTC fee BM

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add check for isActivated and if not return LEGACY_BURNING_MAN_BTC_FEES_ADDRESS

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Increase ISSUANCE_BOOST_FACTOR to 3

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Use HashSet instead Lists.
Remove sorting as not needed.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add DateUtil and Hex.decodeLast4Bytes

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add Config options

HenrikJannsen committed:
Add `String formatCoin(long value)` API to CoinFormatter

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Make AveragePriceUtil.getAveragePriceTuple mor flexible
Add filter for before date

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add support for multiple clients calling the setup method.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add support for multiple clients adding a block handler.
Add requestRawDtoBlock method.
Rename to make different block types more clear.
Change visibility

HenrikJannsen committed:
Add new fields and methods

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add messages and domain data for accounting domain

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add BurningManAccountingService to CorePersistedDataHost

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add BurningMan accounting domain

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add BM account UI

HenrikJannsen committed:
Make codacy happy

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Add BurningManAccountingStore resource file.
Use resourceDataStoreService and extend StoreService

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen commented:
We need to update https://bisq.wiki/Dispute_resolution as well.

HenrikJannsen committed:
Update TAC text and layout

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Make copy icon in popups smaller

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Improve TAC text

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Apply code review feedback

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Improve receiver distribution to avoid that legacy BM gets outputs if some outputs fall below the min. amount limit.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Remove TODO, rename method

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

ghost commented:
Regtest Testing notes:

- Seed node requires a command line option `isBmFullNode=true` to enable accounting.
- DAO genesis tx may need to be generated at block height 111, since the code assumes that.

sqrrm commented on core/src/main/java/bisq/core/dao/burningman/accounting/storage/BurningManAccountingStoreService.java:
```diff
@@ -0,0 +1,92 @@
+        requestPersistence();
+    }
+
+    public void purgeLastTenBlocks() {
```
Why is this needed?

HenrikJannsen replied:
Thats used by applyReOrg(). In case of a reorg we remove the last 10 blocks as we do not know how deep the reorg is, but for sure not > 10 blocks (i guess historically it was max 4)

sqrrm commented on core/src/main/java/bisq/core/dao/burningman/accounting/BurningManAccountingService.java:
```diff
@@ -0,0 +1,257 @@
+        // We use the average 30 day BSQ price from the first day of a month back 30 days. So for 1.Nov 2022 we take the average during October 2022.
+        // Filling the map takes a bit of computation time (about 5 sec), so we use for historical data a pre-calculated list.
+        // Average price from 1. May 2019 (April average) - 1. Nov 2022 (Oct average)
+        String historical = "1648789200000=2735, 1630472400000=3376, 1612155600000=6235, 1559365200000=13139, 1659330000000=3609, 1633064400000=3196, 1583038800000=7578, 1622523600000=3918, 1625115600000=3791, 1667278800000=3794, 1561957200000=10882, 1593579600000=6153, 1577854800000=9034, 1596258000000=6514, 1604206800000=5642, 1643691600000=3021, 1606798800000=4946, 1569906000000=10445, 1567314 ...
```
Line break would be nice

HenrikJannsen replied:
Ok will add. Intellij has also smart linebreaks, breaking at window width.

sqrrm commented on core/src/main/java/bisq/core/dao/burningman/accounting/BurningManAccountingService.java:
```diff
@@ -0,0 +1,257 @@
+        });
+    }
+
+    private Map<Date, Price> getAverageBsqPriceByMonth(Date from, int toYear, int toMonth) {
```
A comment that it's counted backwards would be helpful, ie, `from > Date(toYear, toMonth)`. A natural assumption with time is to think of it as moving forwards.

HenrikJannsen replied:
There is no reason why I used it backwards, but prefer not to change it as it is tested with that already.

sqrrm commented on core/src/main/java/bisq/core/dao/burningman/accounting/balance/BalanceModel.java:
```diff
@@ -0,0 +1,124 @@
+                .collect(Collectors.toList());
+    }
+
+    private Set<Date> getMonths(Date from, int toYear, int toMonth) {
```
Comment on usage would be nice.

HenrikJannsen replied:
Will add. We cannot ensure burnOutputModelsByMonth covers all months, so we create a set of months to iterate over

sqrrm commented on core/src/main/java/bisq/core/dao/burningman/accounting/node/AccountingNodeProvider.java:
```diff
@@ -0,0 +1,57 @@
+                && preferences.getRpcPw() != null &&
+                !preferences.getRpcPw().isEmpty() &&
+                preferences.getBlockNotifyPort() > 0;
+        if (BurningManService.isActivated()) {
+            accountingNode = isBmFullNode && rpcDataSet ? fullNode : liteNode;
```
A warning that it's not running in accounting fullnode mode when config is set to fullnode would be good, in case of rpc data issues.

HenrikJannsen replied:
Ok

sqrrm commented on core/src/main/java/bisq/core/dao/burningman/accounting/node/full/AccountingBlockParser.java:
```diff
@@ -0,0 +1,210 @@
+        // Max fees amount is currently 0.0176 BTC.
+        // We give some extra headroom to cover potential future fee changes.
+        // We store value as integer in protobuf to safe space, so max. possible value is 21.47483647 BTC.
+        long maxTradeFeeValue = Coin.parseCoin("0.1").getValue();
```
The comment says 1_000_000 sats but this is 10_000_000 sats. I think 10m is correct and the comment should be updated.

HenrikJannsen replied:
Ok, will fix

sqrrm commented on core/src/main/java/bisq/core/dao/burningman/accounting/node/full/AccountingBlockParser.java:
```diff
@@ -0,0 +1,210 @@
+        return Optional.empty();
+    }
+
+    // TODO not sure if other ScriptType are to be expected
+    private boolean isExpectedScriptType(TempAccountingTxOutput txOutput, TempAccountingTx accountingTx) {
```
Why is it needed to check for script type at all?

HenrikJannsen replied:
Its just an attempt to narrow options for false positives.

sqrrm commented on core/src/main/java/bisq/core/dao/burningman/accounting/node/full/AccountingFullNode.java:
```diff
@@ -0,0 +1,323 @@
+
+    // We start after initial DAO parsing is complete
+    @Override
+    protected void onInitialDaoBlockParsingComplete() {
```
It makes sense to start this after dao parsing is done, but I feel it's hard to get an understanding for when services are run. Now it's not a big issue with just a couple staggered like this, but as a principle it would probably be better to have a startup handler that did the scheduling for easier overview.

HenrikJannsen replied:
Yes, that got a bit messy over time. But I don't want to do bigger refactorings there.

HenrikJannsen commented on desktop/src/main/java/bisq/desktop/main/dao/burnbsq/burningman/BalanceEntryItem.java:
```diff
@@ -0,0 +1,219 @@
+        if (price.isEmpty() || receivedBtc.isEmpty()) {
+            receivedBtcAsBsq = Optional.empty();
+        } else {
+            long volume = price.get().getVolumeByAmount(Coin.valueOf(receivedBtc.get())).getValue();
```
You mean the receivedBtc.get() is 0 in that case and getVolumeByAmount throws an exception?

HenrikJannsen replied:
Ah I see when the price is 0 the converter methods throw an exception. Unfortunately the ExchangeRate class is from BitcoinJ and lacks a check. I will create a separate PR for moving all those calls to the VolumeUtil and add a check there.

HenrikJannsen replied:
Actually an exception is the correct behaviour for the API here. The caller need to avoid calls with price 0. I will add a check in that code part.

pazza83 commented:
> We need to update https://bisq.wiki/Dispute_resolution as well.

I am happy to do this and also update:

- https://bisq.wiki/Donation_Address_Owner
- https://bisq.wiki/Arbitrator
- https://bisq.wiki/Arbitration

Assuming https://github.com/bisq-network/proposals/issues/397 will also be accepted? If so I can explain this in the wiki.

Also assuming this proposal https://github.com/bisq-network/proposals/issues/386 is also accepted? 

Also unsure as to how legacy trades will be managed?

Will trade placed before the implementation of the new burning men proposal still be able to be taken?

When will https://github.com/bisq-network/proposals/issues/386 be actioned by the refund agent?

Will it apply to ALL trades from a specific date?

Will it apply to trades using the new protocol only or legacy trades?

Might be simpler to choose a cut off date that coincides with the new release. Ie trades taken after X date will not have peers trade deposit paid out.

pazza83 commented:
Also need to update the mediation result text. Currently reads:

```
Mediation result for trade with ID: XXXXXXXXXX

The mediator has suggested the following payout:
You receive: X BTC
Your trading peer receives: X BTC

You can accept or reject this suggested payout.

By accepting, you sign the proposed payout transaction. If your trading peer also accepts and signs, the payout will be completed, and the trade will be closed.

If one or both of you reject the suggestion, you will have to wait until XX Month TIME (block XXX,XXX) to open a second-round dispute with an arbitrator who will investigate the case again and do a payout based on their findings.

The arbitrator may charge a small fee (fee maximum: the trader's security deposit) as compensation for their work. Both traders agreeing to the mediator's suggestion is the happy path—requesting arbitration is meant for exceptional circumstances, such as if a trader is sure the mediator did not make a fair payout suggestion (or if the other peer is unresponsive).

More details about the new arbitration model: [1]
[https://bisq.wiki/Dispute_resolution#Level_3:_Arbitration]
```


Suggested change:

Replace penultimate paragraph with:

`Mediation is expected to be the optimal resolution for both traders. If the trade goes to arbitration the arbitrator will payout out the trade amount plus one peer's security deposit. This means the total arbitration payout will be less than the mediation payout. Both traders agreeing to the mediator's suggestion is the happy path. Requesting arbitration is meant for exceptional circumstances, such as one peer not responding, or if a trader is sure the mediator did not make a fair payout suggestion.`

HenrikJannsen committed:
Remove unused var

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Use constants for opReturn data. Rename hash to opReturnData.
Update UI.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen committed:
Remove DevEnv.isDevTesting
Add check for number of confirmations in devMode to show support button after 5 blocks

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen commented:
Here are some notes for testing with some explainations about certain behaviour:

- It is expected that genesis tx is at block 111. If your setup has it different change the value in BurningManAccountingService.EARLIEST_BLOCK_HEIGHT
- We go back 12 cycles for burn target calculations. To make testing more consistant start by creating 156 blocks (13*12)
- Compensation requests go back 24 cycles until they decay to 0
- For burn target we go back 12 cycles
- At trades we need to ensure that both traders have same block. To guarantee that we use a past snapshot block which is always the same for both traders even one has a newer block or there is fork or reorg.
  The block height is the last mod(10) height from the range of the last 10-20 blocks (139 -> 120; 140 -> 130, 141 -> 130)
- To test the legacy BM one need to burn some BSQ with pre-image `fee` for the fee receiver BM and `dpt` for the delayed payout tx BM. 
- For BTC distribution only the DPT BM is used (that use the default address from the DAO voting)  
- Receiver address for legacy BM is 2MzBNTJDjjXgViKBGnatDU3yWkJ8pJkEg9w, defined in Para.RECIPIENT_BTC_ADDRESS. If you do not use the default setup which comes with the wallet containing that address, you can either edit the code or do a DAO voting to change the address.
- When testing a DPT case you need to create 5 blocks after take offer to show the open dispute button and to make the lock time valid. 
- If you create multiple blocks the lite client might get out of sync. Also a full client could get issues sometimes. For testing a full client is better as it gets the block immediately, the lite client has a few seconds delay which cause problems when creating fast blocks.

Program artuments for localhost/regtest seed node in full BM mode:

```
--useLocalhostForP2P=true
--useDevPrivilegeKeys=true
--isBmFullNode=true
--useDevMode=true
--baseCurrencyNetwork=BTC_REGTEST
--nodePort=2002
--appName=bisq-BTC_REGTEST_Seed_2002
--fullDaoNode=true
--rpcUser=...
--rpcPassword=...
--rpcPort=18443
--rpcBlockNotificationPort=5120
```

The important ones are:
`--useDevPrivilegeKeys=true` - we use dev privileged keys for broadcasting the data. With that flag it use the hard coded dev keys.

`--isBmFullNode=true` - makes the node a full BM node. Publishes new AccountingBlock data and respond to GetAccountingBlocksRequests.

`--useDevMode=true` - Enables testing with DPT.


Parameters:
ISSUANCE_BOOST_FACTOR= 3;
MAX_BURN_SHARE = 0.11
BURN_TARGET_BOOST_AMOUNT = 10 000 BSQ
DEFAULT_ESTIMATED_BTC_TRADE_FEE_REVENUE_PER_CYCLE=  10 000 BSQ
  
Initial state:
Block 298
Nothing burned



Expected: Legacy DPT BM get 100% of receiver share

1. 
Legacy BM and DPT and fee burn BSQ:
Burn BSQ in the Proof of Burn view (not BM view) for with fee and dpt pre-image and check if after a new block the 2 legacy BM in the UI have the burned amounts shown.
As no BSQ got burned so far from contributors, the legacy DPT BM get 100% of receiver share. Legacy BM for fee does not get anything.

Create an offer and after new block check if legacy BM got the trade fee.

Take the offer and check if legacy BM got the del. payout tx after letting the trade go to arbitration.


2.
Burn BSQ as co-founder.
Burn target: 120 000 - 130 000 BSQ
Allowed for a 40% co-founder: 13200 - 14300 BSQ (capped to 11% of 120k-130k)
Burn 13200 BSQ
After a block: 
Burn target:  106800.00 - 116800.00 BSQ (120000-13200)
Receiver share 11%; my new burn target: 0; 
Co-founder 0 (60%) burn target: 1631.46

Why 1631.46?
We have 13200 BSQ burned. Co-founder 0 could get max 11%. If he burns 1631.46 he has 11% of the distribution. Burning more would not gain him more.
13200 + 1631.46 = 14831.46 ->  11% of 14831 = 1631.46 (see getMissingAmountToReachTargetShare method in code)

3.
Let Co-founder 0 burn 1631.46 BSQ.

Both Co-founder 1 and Co-founder 2 have burn target 0 as thye have reached their 11% cap it would not make sense to burn more.


4. 
Alice makes comp request for 20 000 BSQ. She use the address from her Bisq btc wallet as receiver address.
Bob makes comp request for 40 000 BSQ. He does not use a custom receiver address.

After voting Alice can burn 1649.32 - 1693.39 BSQ and Bob: 1693.39 - 1693.39 BSQ .
Burn target is: 105168.54 - 115168.54
Co-founder 0 decayed burn amount: 1516.42
Co-founder 1 decayed burn amount: 12184.62
Alice issuance share: 10.74%
Bob issuance share: 21.49%


Burn target are calculated as follows:
Sum of all decayed burn amounts:
12184.62+1516.42=13701.04
0.11/0.89*13701.04=1693.38 -> If I burn 1693.38 I get 11% of burn share.


Why has Alice less: 
Because her issuance share is < 11% 

5. Alice burn 1649.32
new block

6. Bob burns 1873.73
new block

Bob has now close to 11% receiver share, Alice about 10%. Due to the dynamic behaviour with burn amount decay at each block it is never exactly as it was targeted (11%).

To test receivers for fees we can create an offer and Alice and Bob have 10/11% chance to get the fee. For DPT we need to create 20 blocks to be sure that the snapshot height used for the calculation is higher then the burn tx height.


7. Create 10 compensation requests for 10 new contributors with 9k BSQ and after voting burn max amount each.
After that we have only the co-founder1 with 11% capped share. all others have lower share and can burn more BSQ.
Legacy BM still gets some % as the sum of all burn shares does not reach 100% yet. 
Let everyone burn their max amount and repeat that with new blocks until the legacy BM does not get anymore and until the burn target gets below 0.
It will take some iterations as the max amount is always capped to the amount which makes sense for the individual BM (11% cap).
This situation is only expected at bootstrapping until the BM market is saturated.

8. When creating new block and entering a new cycle we do not expect much change beside the decay because the total burned BSQ is decaying slowly and the added estimated fee revenue of 10k is not triggering a sudden change. This will change once the oldest burn tx falls out of the decay window of 12 cycles. As we likely have created most burn tx in the same cycle we will see a sudden change. Lets repeat to create 13 blocks until we get to that point.

When getting there the oldest burn txs will get decayed value 0 and the legacy BM might get a share again as we might not reach the 100%. In real life situation it is expected that each cycle new burn txs gets added so that effect will likely not happen.
Also the sudden increase of burn target would be avoided then. 
In this test scenario it can be also that the individual burn target gets very low, because the sum of all decayed burn amounts get low as well if we do not burn each cycle...
 
If the sum of all decayed burned BSQ is very low the individual BM might get values below 5.46 BSQ and then its set to 0. 
I will make a change to avoid that by allowing always 1000 BSQ to get burned if on has issuance share and change the UI a bit to show only the suggested value but allow up to 1000 BSQ even if the suggested value is 0 because the BM has already reached its cap.

HenrikJannsen committed:
Add MAX_BURN_TARGET_LOWER_FLOOR to avoid cases where small burn amounts lead to deadlock situations.

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen commented:
> > We need to update https://bisq.wiki/Dispute_resolution as well.
> 
> I am happy to do this and also update:
> 
>     * https://bisq.wiki/Donation_Address_Owner
> 
>     * https://bisq.wiki/Arbitrator
> 
> 

Great thanks!

> Assuming [bisq-network/proposals#397](https://github.com/bisq-network/proposals/issues/397) will also be accepted? If so I can explain this in the wiki.
> 

Seems it got enough support.

> Also assuming this proposal [bisq-network/proposals#386](https://github.com/bisq-network/proposals/issues/386) is also accepted?

Seems it got enough support.

> 
> Also unsure as to how legacy trades will be managed?
> 

The update will happen like that:

Release of 1.9.7 with new BM feature deactivated.
After 1 week after release we set required trade version to 1.9.7. in filter, so anyone who trades need to updated. Pending trades are not affected only new ones.
At activation date (currently planned 1.1.2023) the new BM code gets used. Users who trade have both the new version due the enforcement by the filter.

Pending trades with old version will be treated like before. The Arbitrator UI has still the custom payout option to follow the past rules. Only trades started after activation date are following new rule. Still Arbitrator has flexibility with custom payouts.

Legacy BM will get BTC until activation date. After that it will take a while until new BM have burned BSQ and legacy BM will not receive anything anymore. 

> Will trade placed before the implementation of the new burning men proposal still be able to be taken?

Offers they are not affected. Trades started before are treated as old trades. The new BM distribution is set at take-offer time.

> 
> When will [bisq-network/proposals#386](https://github.com/bisq-network/proposals/issues/386) be actioned by the refund agent?

For trades which started after activation date.

> 
> Will it apply to ALL trades from a specific date?

Yes.

HenrikJannsen commented:
> Also need to update the mediation result text. Currently reads:
> 
> ```
> Mediation result for trade with ID: XXXXXXXXXX
> 
> The mediator has suggested the following payout:
> You receive: X BTC
> Your trading peer receives: X BTC
> 
> You can accept or reject this suggested payout.
> 
> By accepting, you sign the proposed payout transaction. If your trading peer also accepts and signs, the payout will be completed, and the trade will be closed.
> 
> If one or both of you reject the suggestion, you will have to wait until XX Month TIME (block XXX,XXX) to open a second-round dispute with an arbitrator who will investigate the case again and do a payout based on their findings.
> 
> The arbitrator may charge a small fee (fee maximum: the trader's security deposit) as compensation for their work. Both traders agreeing to the mediator's suggestion is the happy path—requesting arbitration is meant for exceptional circumstances, such as if a trader is sure the mediator did not make a fair payout suggestion (or if the other peer is unresponsive).
> 
> More details about the new arbitration model: [1]
> [https://bisq.wiki/Dispute_resolution#Level_3:_Arbitration]
> ```
> 
> Suggested change:
> 
> Replace penultimate paragraph with:
> 
> `Mediation is expected to be the optimal resolution for both traders. If the trade goes to arbitration the arbitrator will payout out the trade amount plus one peer's security deposit. This means the total arbitration payout will be less than the mediation payout. Both traders agreeing to the mediator's suggestion is the happy path. Requesting arbitration is meant for exceptional circumstances, such as one peer not responding, or if a trader is sure the mediator did not make a fair payout suggestion.`



Could you use the full text  as here for your suggestion? Was not totally sure where to insert it.

```
portfolio.pending.mediationResult.popup.info=The mediator has suggested the following payout:\n\
  You receive: {0}\n\
  Your trading peer receives: {1}\n\n\
  You can accept or reject this suggested payout.\n\n\
  By accepting, you sign the proposed payout transaction. \
  If your trading peer also accepts and signs, the payout will be completed, and the trade will be closed.\n\n\
  If one or both of you reject the suggestion, you will have to wait until {2} (block {3}) to open a \
  second-round dispute with an arbitrator who will investigate the case again and do a payout based on their findings.\n\n\
  The arbitrator may charge a small fee (fee maximum: the trader''s security deposit) as compensation for their work. \
  Both traders agreeing to the mediator''s suggestion is the happy pathârequesting arbitration is meant for \
  exceptional circumstances, such as if a trader is sure the mediator did not make a fair payout suggestion \
  (or if the other peer is unresponsive).\n\n\
  More details about the new arbitration model: [HYPERLINK:https://bisq.wiki/Dispute_resolution#Level_3:_Arbitration]
  ```

pazza83 commented:
> Could you use the full text as here for your suggestion? Was not totally sure where to insert it.

No problems:

```
portfolio.pending.mediationResult.popup.info=The mediator has suggested the following payout:\n\
  You receive: {0}\n\
  Your trading peer receives: {1}\n\n\
  You can accept or reject this suggested payout.\n\n\
  By accepting, you sign the proposed payout transaction. \
  Mediation is expected to be the optimal resolution for both traders. \
  If your trading peer also accepts and signs, the payout will be completed, and the trade will be closed.\n\n\
  If one or both of you reject the suggestion, you will have to wait until {2} (block {3}) to reject mediation suggestion, \
  which will open a second-round dispute with an arbitrator who will investigate the case again and do a payout based on their findings.\n\n\
  Mediation is expected to be the optimal resolution for both traders. \
  If the trade goes to arbitration the arbitrator will payout out the trade amount plus one peer's security deposit. \
  This means the total arbitration payout will be less than the mediation payout. \
  Requesting arbitration is meant for exceptional circumstances. such as; \
  one peer not responding, or disputing the mediator made a fair payout suggestion. \n\n\
  More details about the arbitration model: [HYPERLINK:https://bisq.wiki/Dispute_resolution#Level_3:_Arbitration]
```

edited to include @MwithM's feedback.

HenrikJannsen committed:
Update mediation popup text

Signed-off-by: HenrikJannsen <boilingfrog@gmx.com>

HenrikJannsen commented on desktop/src/main/java/bisq/desktop/main/dao/burnbsq/burningman/BurningManView.java:
```diff
@@ -0,0 +1,1540 @@
+                    .map(balanceEntry -> new BalanceEntryItem(balanceEntry, averageBsqPriceByMonth, bsqFormatter, btcFormatter))
+                    .collect(Collectors.toList()));
+            // 108869: 617 - 1878, 640-1531
+            log.error("balanceEntryObservableList setAll took {} ms size={}", System.currentTimeMillis() - ts, balanceEntryObservableList.size());
```
Ups, will remove...

ripcurlx commented:
@jmacxx Could you please ACK the new changes, so I can merge them for the release branch? Thanks!

ghost commented:
Due to BSQ price not having history in regtest, getting IllegalArgumentException (as a popup).
This was previously discussed, when clicking on a burningman candidate row, it still pops up an error box like this:

![image](https://user-images.githubusercontent.com/47253594/207376940-b6a3dbbd-a75d-4635-a561-dba23a4830a1.png)

    Dec-13 09:35:24.817 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: Uncaught Exception from thread JavaFX Application Thread 
    Dec-13 09:35:24.817 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: throwableMessage= null 
    Dec-13 09:35:24.817 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: throwableClass= class java.lang.IllegalArgumentException 
    Dec-13 09:35:24.817 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: Stack trace:
    java.lang.IllegalArgumentException
	    at com.google.common.base.Preconditions.checkArgument(Preconditions.java:130)
	    at bisq.core.monetary.AltcoinExchangeRate.<init>(AltcoinExchangeRate.java:44)
	    at bisq.core.monetary.AltcoinExchangeRate.<init>(AltcoinExchangeRate.java:54)
	    at bisq.core.monetary.Price.getVolumeByAmount(Price.java:87)
	    at bisq.desktop.main.dao.burnbsq.burningman.BalanceEntryItem.<init>(BalanceEntryItem.java:110)
	    at bisq.desktop.main.dao.burnbsq.burningman.BurningManView.lambda$onBurningManSelected$17(BurningManView.java:640)
	    at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	    at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	    at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	    at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	    at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	    at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	    at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	    at bisq.desktop.main.dao.burnbsq.burningman.BurningManView.onBurningManSelected(BurningManView.java:641)

ripcurlx reviewed (approved):
utACK - based on https://github.com/bisq-network/bisq/pull/6423#pullrequestreview-1214722656

ripcurlx commented:
> Due to BSQ price not having history in regtest, getting IllegalArgumentException (as a popup). This was previously discussed, when clicking on a burningman candidate row, it still pops up an error box like this:
> 
> ![image](https://user-images.githubusercontent.com/47253594/207376940-b6a3dbbd-a75d-4635-a561-dba23a4830a1.png)
> 
> ```
> Dec-13 09:35:24.817 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: Uncaught Exception from thread JavaFX Application Thread 
> Dec-13 09:35:24.817 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: throwableMessage= null 
> Dec-13 09:35:24.817 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: throwableClass= class java.lang.IllegalArgumentException 
> Dec-13 09:35:24.817 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: Stack trace:
> java.lang.IllegalArgumentException
>     at com.google.common.base.Preconditions.checkArgument(Preconditions.java:130)
>     at bisq.core.monetary.AltcoinExchangeRate.<init>(AltcoinExchangeRate.java:44)
>     at bisq.core.monetary.AltcoinExchangeRate.<init>(AltcoinExchangeRate.java:54)
>     at bisq.core.monetary.Price.getVolumeByAmount(Price.java:87)
>     at bisq.desktop.main.dao.burnbsq.burningman.BalanceEntryItem.<init>(BalanceEntryItem.java:110)
>     at bisq.desktop.main.dao.burnbsq.burningman.BurningManView.lambda$onBurningManSelected$17(BurningManView.java:640)
>     at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
>     at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
>     at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
>     at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
>     at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
>     at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
>     at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
>     at bisq.desktop.main.dao.burnbsq.burningman.BurningManView.onBurningManSelected(BurningManView.java:641)
> ```

@HenrikJannsen Do we handle this or drop it for now as it won't be a production issue?

HenrikJannsen commented:
> > Due to BSQ price not having history in regtest, getting IllegalArgumentException (as a popup). This was previously discussed, when clicking on a burningman candidate row, it still pops up an error box like this:
> > ![image](https://user-images.githubusercontent.com/47253594/207376940-b6a3dbbd-a75d-4635-a561-dba23a4830a1.png)
> > ```
> > Dec-13 09:35:24.817 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: Uncaught Exception from thread JavaFX Application Thread 
> > Dec-13 09:35:24.817 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: throwableMessage= null 
> > Dec-13 09:35:24.817 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: throwableClass= class java.lang.IllegalArgumentException 
> > Dec-13 09:35:24.817 [JavaFX Application Thread] ERROR bisq.common.setup.CommonSetup: Stack trace:
> > java.lang.IllegalArgumentException
> >     at com.google.common.base.Preconditions.checkArgument(Preconditions.java:130)
> >     at bisq.core.monetary.AltcoinExchangeRate.<init>(AltcoinExchangeRate.java:44)
> >     at bisq.core.monetary.AltcoinExchangeRate.<init>(AltcoinExchangeRate.java:54)
> >     at bisq.core.monetary.Price.getVolumeByAmount(Price.java:87)
> >     at bisq.desktop.main.dao.burnbsq.burningman.BalanceEntryItem.<init>(BalanceEntryItem.java:110)
> >     at bisq.desktop.main.dao.burnbsq.burningman.BurningManView.lambda$onBurningManSelected$17(BurningManView.java:640)
> >     at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
> >     at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
> >     at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
> >     at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
> >     at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
> >     at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
> >     at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
> >     at bisq.desktop.main.dao.burnbsq.burningman.BurningManView.onBurningManSelected(BurningManView.java:641)
> > ```
> 
> @HenrikJannsen Do we handle this or drop it for now as it won't be a production issue?

I fixed it in upcoming PR https://github.com/bisq-network/bisq/pull/6463

ripcurlx commented:
@HenrikJannsen 
I just was about to testing the new distributed burning man.
I created new compensation requests, using once a Bitcoin core address and once the default Bisq wallet.
Now I moved to the Proof of Burn > Burningmen section selecting the contributor I want to use and entering the amount to burn. What is kind of confusing is, that a range is shown that can be used to burn, but it is possible to use a value below as well.

![Bildschirmfoto 2022-12-19 um 10 35 03](https://user-images.githubusercontent.com/170962/208399659-e5af4272-8404-457e-9e2a-11ad9e617147.png)


![Bildschirmfoto 2022-12-19 um 10 36 43](https://user-images.githubusercontent.com/170962/208394552-479f884a-ad35-43ad-856a-265f383b3f41.png)

If there haven't been any CR for this user we should deactivate the dropdown in the `Burn BSQ` section.

![Bildschirmfoto 2022-12-19 um 11 01 34](https://user-images.githubusercontent.com/170962/208399764-2cba4e97-3937-4c91-bc27-485e256a248a.png)

What is the `+` column for?

![Bildschirmfoto 2022-12-19 um 11 24 37](https://user-images.githubusercontent.com/170962/208404317-c9e217ae-46c7-4b0e-854f-22c50368bdf7.png)
I'm not sure if we have this error handling everywhere, but if I use all of my BSQ with a dust output of 0 BSQ I get this error message.

So the general behavior will be that contributors will try to burn as little as possible to reach a maximum receiver share and when slowly more contributors are joining they will increase their burned amount up to max their expected BSQ to receive.

This is one thing that might be a bit confusing is that we display the Expected BSQ to receive, but in reality they will receive BTC, correct?

As this is a feature only for contributors I'm not sure if we should display the `Burningmen` section at all to users that haven't contributed yet.

HenrikJannsen commented:
> @HenrikJannsen I just was about to testing the new distributed burning man. I created new compensation requests, using once a Bitcoin core address and once the default Bisq wallet. Now I moved to the Proof of Burn > Burningmen section selecting the contributor I want to use and entering the amount to burn. What is kind of confusing is, that a range is shown that can be used to burn, but it is possible to use a value below as well.

The `?` at the column header should give a bit of explaination. 
The left part is what amount gives your the max. distribution share based on the already burned BSQ. If there is no BSQ yet burned we use the burn target as base (e.g. if burn target is 100k and you can get max. 11% distribution share it is 11000BSQ).
The right part ist the max burn amount based on the boosted burn target (right side of the total burn target) with the max. distribution share. Distribution share is now 10 x the issuance share or max. 11%.

> 
> If there haven't been any CR for this user we should deactivate the dropdown in the `Burn BSQ` section.

Yes, agree could be added. 

> 
> What is the `+` column for?

To show/hide columns.

 I'm not sure if we have this error handling everywhere, but if I use all of my BSQ with a dust output of 0 BSQ I get this error message.

Hm... yes thats a generic error from the wallet. if you have 10 bsq and want to send 6 bsq you get 4 bsq change thus a dust amount you cannot spend.


> 
> So the general behavior will be that contributors will try to burn as little as possible to reach a maximum receiver share and when slowly more contributors are joining they will increase their burned amount up to max their expected BSQ to receive.

If they act purely economically rational, yet the first BM would start with the smallest amount 5.46 BSQ and get his max. share. Then when the next BM joins his share goes down and he need to burn more.... But thats only in the bootstrapping to get to reasonable amounts. I assume some BM will burn realistic amounts and by that short cut this small incremental process. But later once bootstrapped economically rational behaviour should lead to a competition for the right "costs"/ profit of the model. If BM dont earn enough they will leave or burn less, by that the profit goes up and new ones might join.


> 
> This is one thing that might be a bit confusing is that we display the Expected BSQ to receive, but in reality they will receive BTC, correct?

Yes, thats calculated with the last months average price. As burning is in BSQ I think having BSQ as accounting currency is easier. But yes we could also use BTC as accounting currency but then the burned amount always need to be converted.

> 
> As this is a feature only for contributors I'm not sure if we should display the `Burningmen` section at all to users that haven't contributed yet.

Hm... yes could be considered. But maybe it gives a fake feeling for privacy then? All the data are public anyway so we only show a UI for stuff which is out there already.
